src/truncate.jl

```diff
@@ -1,24 +1,81 @@
 
-The lower bound `l` can be finite or `-Inf` and the upper bound `u` can be finite or
-`Inf`. The function throws an error if `l > u`.
+```math
+f(x; d_0, l, u) = \\frac{f_{d_0}(x)}{P_{Z \\sim d_0}(l \\le Z \\le u)}, \\quad x \\in [l, u],
+```
+where ``f_{d_0}(x)`` is the probability density (mass) function of ``d_0``.
+
+The function throws an error if ``l > u``.
+
+```julia
+truncated(d0; lower=l)           # d0 left-truncated to the interval [l, Inf)
+truncated(d0; upper=u)           # d0 right-truncated to the interval (-Inf, u]
+truncated(d0; lower=l, upper=u)  # d0 truncated to the interval [l, u]
+truncated(d0, l, u)              # d0 truncated to the interval [l, u]
+```
 
 The function falls back to constructing a [`Truncated`](@ref) wrapper.
 
 # Implementation
 
-To implement a specialized truncated form for distributions of type `D`, the method
-`truncated(d::D, l::T, u::T) where {T <: Real}` should be implemented.
+To implement a specialized truncated form for distributions of type `D`, one or more of the
+following methods should be implemented:
+- `truncated(d0::D, l::T, u::T) where {T <: Real}`: interval-truncated
+- `truncated(d0::D, ::Nothing, u::Real)`: right-truncated
+- `truncated(d0::D, l::Real, u::Nothing)`: left-truncated
 """
+function truncated end
 function truncated(d::UnivariateDistribution, l::Real, u::Real)
     return truncated(d, promote(l, u)...)
 end
+function truncated(
+    d::UnivariateDistribution;
+    lower::Union{Real,Nothing}=nothing,
+    upper::Union{Real,Nothing}=nothing,
+)
+    return truncated(d, lower, upper)
+end
+function truncated(d::UnivariateDistribution, ::Nothing, u::Real)
+    # (log)ucdf = (log)tp = (log) P(X ≤ u) where X ~ d
+    logucdf = logtp = logcdf(d, u)
+    ucdf = tp = exp(logucdf)
+
+    Truncated(d, promote(oftype(float(u), -Inf), u, zero(ucdf), ucdf, tp, logtp)...)
+end
+function truncated(d::UnivariateDistribution, l::Real, ::Nothing)
+    # (log)lcdf = (log) P(X < l) where X ~ d
+    loglcdf = if value_support(typeof(d)) === Discrete
+        logsubexp(logcdf(d, l), logpdf(d, l))
+    else
+        logcdf(d, l)
+    end
+    lcdf = exp(loglcdf)
+
+    # (log)tp = (log) P(l ≤ X) where X ∼ d
+    logtp = log1mexp(loglcdf)
+    tp = exp(logtp)
 
+    Truncated(d, promote(l, oftype(float(l), Inf), lcdf, one(lcdf), tp, logtp)...)
+end
+truncated(d::UnivariateDistribution, ::Nothing, ::Nothing) = d
 function truncated(d::UnivariateDistribution, l::T, u::T) where {T <: Real}
     l <= u || error("the lower bound must be less or equal than the upper bound")
+    force_kwarg = VERSION ≥ v"1.5.2" ? (force=true,) : ()
+    l == -Inf && Base.depwarn(
+        "`truncated(d, -Inf, u)` is deprecated. Please use `truncated(d; upper=u)` instead.",
+        :truncated;
+        force_kwarg...
+    )
+    u == Inf && Base.depwarn(
+        "`truncated(d, l, Inf)` is deprecated. Please use `truncated(d; lower=l)` instead.",
+        :truncated;
+        force_kwarg...
+    )
```

Comment:
Do we really want to force the deprecation warnings? This could affect performance quite severely, e.g., typically sampling with Turing is basically impossible if there are deprecation in the model. Since deprecation warnings are only enabled in tests by default in recent Julia versions, it will be less disruptive if we don't use `force`. Usually (e.g., also for `Truncated`) we use regular deprecations.

Reply:
I was thinking in this case it's mostly end users who will need to update their code, not package maintainers. e.g. a lot of Turing models and tutorials. If we don't force the deprecation warning, then likely very few of those users will be aware of the change before the next breaking release. I'm not so much concerned about performance, since all they have to do is switch to the new recommended syntax to avoid the deprecation warnings entirely. And the forced warning will fill their REPL pretty quickly, so I doubt there will be many hold-outs.

But maybe this is not a special case, and there's no point deviating from the norm in this package.

Reply:
I can see the possible benefits but they would apply to many other deprecations, also in other packages. The default, however, in this package and the `@deprecate` macro in general is to not enforce deprecation warnings. And I don't think there's a completely convincing reason, at least not for me, why we should deviate from this behaviour here (but e.g. not when deprecating `Truncated`).

Reply:
Okay, I removed the forcing.